exporter/elasticsearchexporter/internal/objmodel/objmodel.go

```diff
@@ -0,0 +1,412 @@
+func DocumentFromAttributes(am pdata.AttributeMap) Document {
+	return DocumentFromAttributesWithPath("", am)
+}
+
+// DocumentFromAttributesWithPath creates a document from a OpenTelemetry attribute
+// map. All nested maps will be flattened, with keys being joined using a `.` symbol.
+//
+// All keys in the map will be prefixed with path.
+func DocumentFromAttributesWithPath(path string, am pdata.AttributeMap) Document {
+	if am.Len() == 0 {
+		return Document{}
+	}
+
+	fields := make([]field, 0, am.Len())
+	fields = appendAttributeFields(fields, path, am)
+	return Document{fields}
+}
+
+// AddTimestamp adds a raw timestamp value to the Document.
+func (doc *Document) AddTimestamp(key string, ts pdata.Timestamp) {
+	doc.Add(key, TimestampValue(ts.AsTime()))
+}
+
+// Add adds a converted value to the document.
+func (doc *Document) Add(key string, v Value) {
+	doc.fields = append(doc.fields, field{key: key, value: v})
+}
+
+// AddString adds a string to the document.
+func (doc *Document) AddString(key string, v string) {
+	if v != "" {
+		doc.Add(key, StringValue(v))
+	}
+}
+
+// AddID adds the hex presentation of an id value to the document. If the ID
+// is empty, no value will be added.
+func (doc *Document) AddID(key string, id idValue) {
+	if !id.IsEmpty() {
+		doc.AddString(key, id.HexString())
+	}
+}
+
+// AddInt adds an integer value to the document.
+func (doc *Document) AddInt(key string, value int64) {
+	doc.Add(key, IntValue(value))
+}
+
+// AddAttributes expands and flattens all key-value pairs from the input attribute map into
+// the document.
+func (doc *Document) AddAttributes(key string, attributes pdata.AttributeMap) {
+	doc.fields = appendAttributeFields(doc.fields, key, attributes)
+}
+
+// AddAttribute converts and adds a AttributeValue to the document. If the attribute represents a map,
+// the fields will be flattened.
+func (doc *Document) AddAttribute(key string, attribute pdata.AttributeValue) {
+	switch attribute.Type() {
+	case pdata.AttributeValueNULL:
+		// do not add 'null'
+	case pdata.AttributeValueMAP:
+		doc.AddAttributes(key, attribute.MapVal())
+	default:
+		doc.Add(key, ValueFromAttribute(attribute))
+	}
+}
+
+// Sort sorts all fields in the document by key name.
+func (doc *Document) Sort() {
+	sort.SliceStable(doc.fields, func(i, j int) bool {
+		return doc.fields[i].key < doc.fields[j].key
+	})
+
+	for i := range doc.fields {
+		fld := &doc.fields[i]
+		fld.value.Sort()
+	}
+}
+
+// Dedup removes fields from the document, that have duplicate keys.
+// The filtering only keeps the last value for a key.
+//
+// NOTE: The document MUST be sorted before dedup can be applied.
+func (doc *Document) Dedup() {
+	// 1. rename fields if a primitive value is overwritten by an object.
+	//    For example the pair (path.x=1, path.x.a="test") becomes:
+	//    (path.x.value=1, path.x.a="test").
+	//
+	//    This step removes potential conflicts when dedotting and serializing fields.
```

Comment:
it looks like this might get confused by `path.x=1, path.x.value="test"`? I'm not sure what the intended behavior is in that case -- it looks like maybe that would just be treated the same as a regular duplicate, which might be unexpected? Whatever the right thing is, it'd be nice if the comment clarified the intent in that case.

Reply:
Your observation is correct. Renaming `path.x` to `path.x.value` is an edge case, yet we've seen this renaming is required from time to time. We have seen this kind of need with docker and k8s labels that have been added to the event as a dictonary.

We could have dropped `path.x`, as it is a conflict/duplicate already. But I did try to keep the logic more similar to Beats here. If we have a conflict on `path.x`, in Beats we rename the field to `path.x.value`, so we can still index it. The rename is required, because ES dedots the fields before indexing.

A second `path.x.value` already in the event would need to be handled as duplicate (edge case on edge case). We just do what we originally would have done, remove the original `path.x`.

Reply:
Added code comments + unit test to validate the expected behavior.